Revert usage of NJT views in SDPA

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #123215

For internal purposes, this PR reverts the use of real views in SDPA -> autograd.Function "views" (i.e. `ViewBufferFromNested` and `ViewNestedFromBuffer`). This is a temporary fix to get the FIRST model launched and working.

**Note: this breaks some other Dynamo tests related to SDPA that rely on real views, but the breakage there isn't expected to be likely in a real-world scenario.**

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

Differential Revision: [D55766281](https://our.internmc.facebook.com/intern/diff/D55766281)

jbschlosser committed:
Revert usage of NJT views in SDPA

[ghstack-poisoned]

YuqingJ commented on torch/nested/_internal/sdpa.py:
Do we need to change line 651 too?

jbschlosser replied:
Not for the purposes of this hotfix. That path is only taken by models like SAM that utilize ragged batch dimensions with fixed sequence lengths. But we should fix it in general. I wanted to keep the scope of this PR limited to the immediate needs of FIRST.

jbschlosser commented:
@pytorchbot merge

YuqingJ commented:
@pytorchbot merge

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your PR but I think it needs to be skipped on ROCm https://hud.pytorch.org/pytorch/pytorch/commit/0fcddb56252c9b4401e8b888eddd4bc4bce3e624' -c nosignal

jithunnair-amd commented:
> * [rocm / linux-focal-rocm6.0-py3.8 / test (default, 3, 6, linux.rocm.gpu.2)](https://hud.pytorch.org/pr/pytorch/pytorch/123215#23424183969) ([gh](https://github.com/pytorch/pytorch/actions/runs/8548621906/job/23424183969))
>   `test_nestedtensor.py::TestNestedTensorSubclassCUDA::test_sdpa_backwards_cuda_bfloat16`

@jbschlosser @YuqingJ This is the specific failure on ROCm:
[rocm / linux-focal-rocm6.0-py3.8 / test (default, 3, 6, linux.rocm.gpu.2)](https://hud.pytorch.org/pr/pytorch/pytorch/123215#23424183969) ([gh](https://github.com/pytorch/pytorch/actions/runs/8548621906/job/23424183969))
test_nestedtensor.py::TestNestedTensorSubclassCUDA::test_sdpa_backwards_cuda_bfloat16

jithunnair-amd commented:
@xinyazhang @groenenboomj for awareness about this SDPA update

jbschlosser committed:
Update on "Revert usage of NJT views in SDPA"


For internal purposes, this PR reverts the use of real views in SDPA -> autograd.Function "views" (i.e. `ViewBufferFromNested` and `ViewNestedFromBuffer`). This is a temporary fix to get the FIRST model launched and working.

**Note: this breaks some other Dynamo tests related to SDPA that rely on real views, but the breakage there isn't expected to be likely in a real-world scenario.**

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge -i

groenenboomj commented:
> cc @xinyazhang @groenenboomj for awareness about this SDPA update

For posterity, this will be in the next PR pass.  Currently finalizing the implementation.

xinyazhang commented:
> test_nestedtensor.py::TestNestedTensorSubclassCUDA::test_sdpa_backwards_cuda_bfloat16

This requires varlen support in ROCM's FA implementation. Before the varlen support is merged, nested tensor related FA UTs should be disabled on ROCM.

jbschlosser commented:
Thanks @groenenboomj / @xinyazhang. I'm skipping this test in ROCm in the meantime.
